### PR TITLE
MINOR: fix generic type of ProcessorParameters

### PR DESCRIPTION
**Pull request by vvcephei**
In unrelated recent work, I noticed some warnings about the missing
type parameters on ProcessorParameters.

While investigating it, it seems like there was a bug in the creation of repartition topics.

### Committer Checklist (excluded from commit message)
- [ ] Verify design and implementation 
- [ ] Verify test coverage and CI build status
- [ ] Verify documentation (including upgrade notes)

### CODE REVIEW

File: streams/src/main/java/org/apache/kafka/streams/kstream/internals/GroupedStreamAggregateBuilder.java
Comment (vvcephei) [inline streams/src/main/java/org/apache/kafka/streams/kstream/internals/GroupedStreamAggregateBuilder.java:87]: Processors are always generically typed with their *input* `K` and `V` types. This was incorrectly using the type of the aggregate as the input value type.

File: streams/src/main/java/org/apache/kafka/streams/kstream/internals/KGroupedTableImpl.java
Comment (vvcephei) [inline streams/src/main/java/org/apache/kafka/streams/kstream/internals/KGroupedTableImpl.java:50]: This is an internal class, and the field is never referenced outside of this class. I think it can be private.

File: streams/src/main/java/org/apache/kafka/streams/kstream/internals/KGroupedTableImpl.java
Comment (vvcephei) [inline streams/src/main/java/org/apache/kafka/streams/kstream/internals/KGroupedTableImpl.java:112]: This chunk is just adding the missing type and reformatting.

File: streams/src/main/java/org/apache/kafka/streams/kstream/internals/KGroupedTableImpl.java
Comment (vvcephei) [inline streams/src/main/java/org/apache/kafka/streams/kstream/internals/KGroupedTableImpl.java:171]: reformatting

File: streams/src/main/java/org/apache/kafka/streams/kstream/internals/KStreamImpl.java
Comment (vvcephei) [inline streams/src/main/java/org/apache/kafka/streams/kstream/internals/KStreamImpl.java]: adding the missing types

File: streams/src/main/java/org/apache/kafka/streams/kstream/internals/KStreamImpl.java
Comment (vvcephei) [inline streams/src/main/java/org/apache/kafka/streams/kstream/internals/KStreamImpl.java:454]: This was incorrectly typed with the result k/v instead of the input k/v

File: streams/src/main/java/org/apache/kafka/streams/kstream/internals/KStreamImpl.java
Comment (vvcephei) [inline streams/src/main/java/org/apache/kafka/streams/kstream/internals/KStreamImpl.java:576]: I think this was a bug. `repartitionForJoin` actually ignored the `otherValueSerde` and only used the `valueSerde`. In this case, we would use the incorrect serde to de/serialize the repartition topic.

I've clarified it by adding the correct generic type parameters and expanding the argument list of `repartitionForJoin` to only take the information it will actually use.

File: streams/src/main/java/org/apache/kafka/streams/kstream/internals/KStreamImpl.java
Comment (vvcephei) [inline streams/src/main/java/org/apache/kafka/streams/kstream/internals/KStreamImpl.java:558]: Renamed these to match the other names used in this class

File: streams/src/test/java/org/apache/kafka/streams/kstream/internals/graph/GraphGraceSearchUtilTest.java
Comment (vvcephei) [inline streams/src/test/java/org/apache/kafka/streams/kstream/internals/graph/GraphGraceSearchUtilTest.java:51]: Since I've added the missing generic parameters to the ProcessorParameters constructor args, it detected that Object/Object was not the right parameterization here.

**Review by vvcephei — COMMENTED** [review]
@guozhangwang @mjsax @bbejeck 

I noticed some warnings around the generics on ProcessorParameters. While investigating it, I think I found a bug in repartitioning.

File: streams/src/main/java/org/apache/kafka/streams/kstream/internals/graph/StatefulProcessorNode.java
Comment (guozhangwang) [inline streams/src/main/java/org/apache/kafka/streams/kstream/internals/graph/StatefulProcessorNode.java:78]: Good catch!

File: streams/src/main/java/org/apache/kafka/streams/kstream/internals/KStreamImpl.java
Comment (bbejeck) [inline streams/src/main/java/org/apache/kafka/streams/kstream/internals/KStreamImpl.java]: nit:  `repartitionNameOverride` won't ever be `null` from doing the checks for the name on lines 571 and 576, so we can get rid of this line

File: streams/src/main/java/org/apache/kafka/streams/kstream/internals/KStreamImpl.java
Comment (guozhangwang) [reply streams/src/main/java/org/apache/kafka/streams/kstream/internals/KStreamImpl.java:558]: Should the `<V1, R> KStream<K, R> outerJoin` and `<V1, R> KStream<K, R> join` above be changed as well?

File: streams/src/main/java/org/apache/kafka/streams/kstream/internals/KStreamImpl.java
Comment (bbejeck) [reply streams/src/main/java/org/apache/kafka/streams/kstream/internals/KStreamImpl.java:576]: Good catch, but I still prefer to use `Joined` as a single parameter, but overall I think this approach is better.

In the Serde inheritance PR the order was switched when creating a new `Joined` object ie `Joined.with(keySerde, otherValueSerde, valueSerde)` so the `otherValueSerde` was used,  but the change was subtle and I missed that point.  Probably better to keep it this way so things are more explicit.

Comment (guozhangwang) [reply streams/src/main/java/org/apache/kafka/streams/kstream/internals/KStreamImpl.java:576]: I agree, actually before @bbejeck 's PR the line was 

```
joinOther.repartitionForJoin(Joined.with(joined.keySerde(), joined.otherValueSerde(), joined.valueSerde()));
```

to intentionally switch other value serde and value serde, but the pervious PR dropped this.

**Review by bbejeck — COMMENTED** [review]
Thanks @vvcephei left some comments

File: streams/src/main/java/org/apache/kafka/streams/kstream/internals/KStreamImpl.java
Comment (guozhangwang) [inline streams/src/main/java/org/apache/kafka/streams/kstream/internals/KStreamImpl.java]: All the callers seems already handles the `name == null` case so this seems unnecessary.

**Review by guozhangwang — COMMENTED** [review]
Made a pass, mostly are minor comments.

File: streams/src/main/java/org/apache/kafka/streams/kstream/internals/KStreamImpl.java
Comment (vvcephei) [reply streams/src/main/java/org/apache/kafka/streams/kstream/internals/KStreamImpl.java]: Thanks for the feedback, but I think it would be null in the call from https://github.com/apache/kafka/pull/5741/files#diff-5142e1d4a6410459d6bf6df98828e5afL682 and https://github.com/apache/kafka/pull/5741/files#diff-5142e1d4a6410459d6bf6df98828e5afL703

Comment (bbejeck) [reply streams/src/main/java/org/apache/kafka/streams/kstream/internals/KStreamImpl.java]: Yes that's correct as this PR stands now.   
But if we put the name check back to what it was originally, then this line is not needed.

File: streams/src/main/java/org/apache/kafka/streams/kstream/internals/KStreamImpl.java
Comment (vvcephei) [reply streams/src/main/java/org/apache/kafka/streams/kstream/internals/KStreamImpl.java:576]: Thanks, all. I'm sorry, but I'm confused about what you're both suggesting to do.
It sounds like this was a regression during refactoring, which makes sense (and makes me feel better that this bug wasn't released).

But are you suggesting to go back to just the `Joined` parameter for `repartitionForJoin`? Or are you ok with this proposal to have two params "repartition name" and "repartition serde"?

Personally, I'm weakly in favor of the current implementation in my PR, since:
* it doesn't require the construction of a new Joined that will just get discarded
* in particular, Joined passes 3 pieces of information, but we only need 2 of them,
* it also allows us to use the type system to prevent another regression of this nature in the future

File: streams/src/main/java/org/apache/kafka/streams/kstream/internals/KStreamImpl.java
Comment (vvcephei) [reply streams/src/main/java/org/apache/kafka/streams/kstream/internals/KStreamImpl.java]: This just simplifies the callers, who would just have to do the same check at the call site: https://github.com/apache/kafka/pull/5741/files#diff-5142e1d4a6410459d6bf6df98828e5afL682 and https://github.com/apache/kafka/pull/5741/files#diff-5142e1d4a6410459d6bf6df98828e5afL703

Comment (vvcephei) [reply streams/src/main/java/org/apache/kafka/streams/kstream/internals/KStreamImpl.java]: Ok, I've pushed the check back out to the callers.

File: streams/src/main/java/org/apache/kafka/streams/kstream/internals/KStreamImpl.java
Comment (vvcephei) [reply streams/src/main/java/org/apache/kafka/streams/kstream/internals/KStreamImpl.java:558]: In response to this comment, I've gone through the whole `KStreamImpl` and standardized the generic types in the following way:
* `K,V` (unchanged) the key/value types of `this` stream
* `KR/VR` the "result" key/value types where applicable
* `KO/VO` the "other" key/value types (where there's an "other" stream involved)
* `KG/VG` the key/value types when there's a global table involved
* `K1/V1` key/value types with no particular semantics that are scoped to the method (I.e., not the same as the instance-level `this` `K/V`)

File: streams/src/main/java/org/apache/kafka/streams/kstream/internals/KStreamImpl.java
Comment (vvcephei) [reply streams/src/main/java/org/apache/kafka/streams/kstream/internals/KStreamImpl.java]: ok, I had removed that check from half the callers, but at Bill's request, I've pushed it back out.

File: streams/src/main/java/org/apache/kafka/streams/kstream/internals/KStreamImpl.java
Comment (bbejeck) [reply streams/src/main/java/org/apache/kafka/streams/kstream/internals/KStreamImpl.java:576]: My apologies for not being clear before. 

I agree with you that I'm in favor of the approach you have here as it's more clear what is going on.  
My original point was I favor the conciseness of a single `Joined` parameter, but it's easier to create a regression (which my PR did), so I vote to pass three parameters to `repartitionForJoin`.

**Review by bbejeck — APPROVED** [review]
Changes LGTM, provided tests pass

File: streams/src/main/java/org/apache/kafka/streams/kstream/internals/KStreamImpl.java
Comment (vvcephei) [reply streams/src/main/java/org/apache/kafka/streams/kstream/internals/KStreamImpl.java:576]: Ah, yes, then I agree on both points.

Comment (guozhangwang) [reply streams/src/main/java/org/apache/kafka/streams/kstream/internals/KStreamImpl.java:576]: > I agree, actually before @bbejeck 's PR the line was

Just clarify the `I agree` was addressing the `I think it is a bug` comment (I think @bbejeck and I were commenting at the same time so these our first two comments were actually in parallel :P). With `Joined` we basically need to do what originally in this code, i.e. creating a newJoined object with the serdes switched. I'm fine with just passing in parameters directly but I'll leave that to your call.

**Comment by mjsax** [comment]
Just skipped over it. Nice cleanup. I often argue with people to rename generic types, but often I get ignored. I'll remember this PR to justify my nitty-ness (is this a word?) in the future.

**Comment by vvcephei** [comment]
Yeah, on one hand, it doesn't matter much, but on the other, I find myself confused fairly often looking at some variable of type T, wondering what the heck it is...

**Comment by guozhangwang** [comment]
Cherry-picked to 2.1 as well.

Thanks @vvcephei !!

**Comment by vvcephei** [comment]
Thank you, @guozhangwang !

**Pull request merged**
